### PR TITLE
Added grover loss functions

### PR DESCRIPTION
**Pull request by gusty1g**
# Pull Request Template

## Description

I am adding grover loss function in this pull request.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/losses.py [hunk 1]
```diff
@@ -465,6 +468,130 @@ def loss(inputs):
...
+            # av_task_dist_loss = nn.KLDivLoss(reduction="mean")
+            av_task_dist_loss = nn.MSELoss(reduction="mean")
+            fg_task_dist_loss = nn.MSELoss(reduction="mean")
+            sigmoid = nn.Sigmoid()
+            av_atom_loss, av_bond_loss, av_dist_loss = 0.0, 0.0, 0.0
+            fg_atom_from_atom_loss, fg_atom_from_bond_loss, fg_atom_dist_loss = 0.0, 0.0, 0.0
+            bv_atom_loss, bv_bond_loss, bv_dist_loss = 0.0, 0.0, 0.0
+            fg_bond_from_atom_loss, fg_bond_from_bond_loss, fg_bond_dist_loss = 0.0, 0.0, 0.0
+
+            if preds["av_task"][0] is not None:
```
Comment (rbharath) [inline deepchem/models/losses.py:508]: Instead of assuming specific structure on `preds`, I'd suggest just accepting different tensors for the different sub-loss terms.

Comment (rbharath) [reply deepchem/models/losses.py:508]: For example, you can have atom, bond, functional group terms

File: deepchem/models/losses.py [hunk 2]
```diff
@@ -465,6 +468,130 @@ def loss(inputs):
         return loss
 
 
+class GroverPretrainLoss(Loss):
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/models/losses.py:471]: We should also rename this loss to something more general (AtomBondFunctionalGroupLoss?)

[the remark this answers is not in the document]
Comment (gusty1g) [reply deepchem/models/losses.py:471]: > Can you explain here the mathematical idea of this loss? There's 3 different comparisons to atoms/bonds/functional groups in this loss. That's a general idea that might be useful to other models besides grover

The overall loss here is the sum of loss from three different model's predictions. We don't compare atoms/bonds/functional groups loss. We compute the individual losses and then sum it to get overall loss.

[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/models/losses.py:471]: New comments are helpful, but let's rename this to something more general since other models may want to use this loss

File: deepchem/models/losses.py [hunk 3]
```diff
@@ -465,6 +468,119 @@ def loss(inputs):
...
+            """
+            av_task_loss = nn.NLLLoss(ignore_index=0,
+                                      reduction="mean")  # same for av and bv
+            fg_task_loss = nn.BCEWithLogitsLoss(reduction="mean")
+            av_task_dist_loss = nn.MSELoss(reduction="mean")
+            fg_task_dist_loss = nn.MSELoss(reduction="mean")
+
+            sigmoid = nn.Sigmoid()
+
+            av_atom_loss = av_task_loss(preds['av_task'][0], targets["av_task"])
```
Comment (rbharath) [inline deepchem/models/losses.py:515]: What is the expected structure of `preds`? Why do we expect it to have a field 'av_task'? We can do one round of simplification here to directly accept atom losses, bond loses, etc as the arguments to this loss function

File: deepchem/models/losses.py [hunk 4]
```diff
@@ -465,6 +468,119 @@ def loss(inputs):
...
+        # TODO It should be BCEWithLogitsLoss if dataset type is classification.
+        # We are assuming here task type to be regression.
+        import torch.nn as nn
+
+        def loss(preds, targets, dist_coff=0.1):
+            pred_loss = nn.MSELoss()
+
+            # print(type(preds))
+            # TODO: Here, should we need to involve the model status? Using len(preds) is just a hack.
+            if type(preds) is not tuple:
```
Comment (rbharath) [inline deepchem/models/losses.py:570]: What is the expected structure of `preds` here? We shouldn't make use of undocumented structure otherwise will be hard to maintain

File: deepchem/models/losses.py [hunk 5]
```diff
@@ -465,6 +468,130 @@ def loss(inputs):
...
+                           dist_coff * bv_dist_loss + fg_dist_loss
+
+            # return overall_loss, av_loss, bv_loss, fg_loss, av_dist_loss, bv_dist_loss, fg_dist_loss
+            # We just return overall_loss since TorchModel can handle only a single loss
+            return overall_loss
+
+        return loss
+
+
+class GroverFinetuneLoss(Loss):
```
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/models/losses.py:558]: Quick ping on this question. Is this loss something other models may want to use?

File: deepchem/models/losses.py [hunk 6]
```diff
@@ -465,6 +468,119 @@ def loss(inputs):
...
+
+    The Grover Pretraining consists learning of atom embeddings and bond embeddings for
+    a molecule. To this end, the learning consists of three tasks:
+        1. Learning of atom vocabulary from atom embeddings and bond embeddings
+        2. Learning of bond vocabulary from atom embeddings and bond embeddings
+        3. Learning to predict functional groups from atom embedings readout and bond embeddings readout
+    The loss function accepts atom vocabulary labels, bond vocabulary labels and functional group
+    predictions produced by Grover model during pretraining as a dictionary and applies negative
+    log-likelihood loss for atom vocabulary and bond vocabulary predictions and Binary Cross Entropy
+    loss for functional group prediction and sums these to get overall loss.
```
Comment (rbharath) [inline deepchem/models/losses.py:482]: Can you add a usage example? Can you add a reference to the grover paper?

File: deepchem/models/losses.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gusty1g) [reply deepchem/models/losses.py:558]: No, other models will not want to use it. It takes two embeddings and make a prediction from it.

File: deepchem/models/losses.py (hunk 3 above)
Comment (gusty1g) [reply deepchem/models/losses.py:515]: There are 3 sets of predictions from GroverPretrain model and they are encoded in the dict `preds`. It is a dict with keys `av_task`, `bv_task` and `fg_task`.

File: deepchem/models/losses.py (hunk 4 above)
Comment (gusty1g) [reply deepchem/models/losses.py:570]: The finetune model in training mode produces a set of output and in evaluation model it produces a different set of outputs. So, it can be a tuple or just the predictions.

File: deepchem/models/losses.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/models/losses.py:558]: Let's chat offline about this one since I want to make sure I understand the loss more precisely

File: deepchem/models/losses.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rbharath) [reply deepchem/models/losses.py:765]: Reading this again, I think it's fine to keep this as `GroverPretrainLoss`. Other models can still use this loss with the context of the new explanation

File: deepchem/models/torch_models/grover_layers.py [hunk 7]
```diff
@@ -262,6 +262,8 @@ def forward(self, embeddings: Dict, atom_scope: List, bond_scope: List):
         """
         preds_atom_from_atom, preds_atom_from_bond, preds_bond_from_atom, preds_bond_from_bond = None, None, None, None
 
+        # TODO remove if condition
+        # see https://github.com/deepchem/deepchem/pull/3300
```
Comment (rbharath) [inline deepchem/models/torch_models/grover_layers.py:266]: Should this be removed since #3300 has been merged in?

File: deepchem/models/losses.py [hunk 8]
```diff
@@ -759,6 +762,133 @@ def log_sum_exp(x, axis=None):
     return y
 
 
+class GroverPretrainLoss(Loss):
```
Comment (rbharath) [inline deepchem/models/losses.py:765]: We need a unit tests for this loss

File: deepchem/models/losses.py [hunk 9]
```diff
@@ -759,6 +762,133 @@ def log_sum_exp(x, axis=None):
...
+    """GroverPretrainLoss
```
Comment (rbharath) [inline deepchem/models/losses.py:766]: A quick reminder to add the loss to the docs

**Review by rbharath — COMMENTED** [review]
A few simple comments below (need to add to docs, need unit test), and I think you have a few TODOs left as well. Let's try to get this merged in by tomorrow so we can get the main Grover PR in

File: deepchem/models/torch_models/grover_layers.py (hunk 7 above)
Comment (gusty1g) [reply deepchem/models/torch_models/grover_layers.py:266]: Its a leftover cleanup from the earlier PR which I would like to do it in a different pull request.

File: deepchem/models/losses.py (hunk 8 above)
Comment (gusty1g) [reply deepchem/models/losses.py:765]: added unit test in 0192870

File: deepchem/models/losses.py (hunk 9 above)
Comment (gusty1g) [reply deepchem/models/losses.py:766]: added to docs in 0192870

**Review by rbharath — APPROVED** [review]
LGTM, feel free to merge once CI is clear
